modules/core/js/src/main/scala/AttributeValue.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 Fabio Labella
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dynosaur
+
+import scala.scalajs.js
+import scala.scalajs.js.typedarray.Uint8Array
+import scala.scalajs.js.annotation._
+
+@js.native
+@JSImport("@aws-sdk/client-dynamodb", "AttributeValue")
+class AttributeValue extends js.Object {
```

Comment:
I'm not sure if this should be a `class` instead of a `trait`. Do you have a reference to the API you are facading?

Reply:
```
export type AttributeValue =
  | AttributeValue.BMember
  | AttributeValue.BOOLMember
  | AttributeValue.BSMember
  | AttributeValue.LMember
  | AttributeValue.MMember
  | AttributeValue.NMember
  | AttributeValue.NSMember
  | AttributeValue.NULLMember
  | AttributeValue.SMember
  | AttributeValue.SSMember
  | AttributeValue.$UnknownMember;

export namespace AttributeValue {
  /**
   * <p>An attribute of type String. For example:</p>
   *         <p>
   *             <code>"S": "Hello"</code>
   *          </p>
   */
  export interface SMember {
    S: string;
    N?: never;
    B?: never;
    SS?: never;
    NS?: never;
    BS?: never;
    M?: never;
    L?: never;
    NULL?: never;
    BOOL?: never;
    $unknown?: never;
  }

  /**
   * <p>An attribute of type Number. For example:</p>
   *         <p>
   *             <code>"N": "123.45"</code>
   *          </p>
   *         <p>Numbers are sent across the network to DynamoDB as strings, to maximize compatibility
   *             across languages and libraries. However, DynamoDB treats them as number type attributes
   *             for mathematical operations.</p>
   */
  export interface NMember {
    S?: never;
    N: string;
    B?: never;
    SS?: never;
    NS?: never;
    BS?: never;
    M?: never;
    L?: never;
    NULL?: never;
    BOOL?: never;
    $unknown?: never;
  }
  
  ...others XMembers
  }
```

On the JS side is nothing more than a JS object with `S`, 'N`, 'B`... keys

I have used `class`. because `@JSImport` requires to have a class. However, I think I shouldn’t need `@JSImport` at all because there is nothing to import I suppose

Reply:
> Besides this, is there any reason you need the npm dependency
At this point, no because this is the only thing we were import

> Btw, I'm curious your of plans to use this library in Scala.js?
The original plan of this library was to offer nice interoperability with the Java SDK. This is why there is no internal ADT for the AttributeValue, but the DynamoValue works as a proxy for the AttributeValue

We have a project written in Scala/JVM that heavily uses this library and the AWS DynamoDb SDK v2. We deploy it as AWS lambda, and we were curious about using feral. I have tried it, and the performances are like 10x the Scala/JVM solution. Adopting feral for the project is blocked by having a ScalaJs Dinosaur release.

On the matter, I would prefer not to expose mutable fields, because the DynamoValue, which is supposed to be immutable, exposes the underlying AttributeValue. Would it be possible to do something like:

```
// Does it need to have the js.native annotation?
trait AttributeValue extends js.Object {
  def S: js.UndefOr[String] = js.native
  def N: js.UndefOr[String] = js.native
  def B: js.UndefOr[Uint8Array] = js.native
  ...
}

object AttributeValue {
  def S(value: String): AttributeValue =
    js.Dynamic.literal(S = value).asInstanceOf[AttributeValue]
  def N(value: String): AttributeValue =
    js.Dynamic.literal(N = value).asInstanceOf[AttributeValue]
  def B(value: Uint8Array): AttributeValue =
    js.Dynamic.literal(B = value).asInstanceOf[AttributeValue]
  ...
}
```

Reply:
> I have tried it, and the performances are like 10x the Scala/JVM solution.

Wow, awesome!! Thanks for sharing.

> Adopting feral for the project is blocked by having a ScalaJs Dinosaur release.

That's exactly how I stumbled here :) I've been looking to rework the DynamoDb integration in Feral and I would really like it to integrate with Dynosaur, and I saw your PR.

Unfortunately, what you have here currently wouldn't integrate with what's currently in Feral. This is because Feral decodes the raw JS objects, first into circe `Json` and then into actual Scala classes.

Indeed, we have our own `AttributeValue` ADT, that is intended to be shared across JVM, JS, and soon Scala Native.
https://github.com/typelevel/feral/blob/309b9fbee0aa3a56f72dfa813723daa500adddc2/lambda/shared/src/main/scala/feral/lambda/events/DynamoDbStreamEvent.scala#L23-L60

So I'm not yet entirely sure what we want to do. One lazy solution is to fork Dynosaur and replace the AWS JVM SDK types with pure Scala classes and cross-compile for the three platforms.

Feel free to open an issue about this on Feral.

Reply:
> On the matter, I would prefer not to expose mutable fields, because the DynamoValue, which is supposed to be immutable, exposes the underlying AttributeValue. Would it be possible to do something like:

Isn't `AttributeValue` mutable on the JVM? How is this any different?

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/model/AttributeValue.html

Reply:
Nope, that is the v1 of the SDK, on the v2, the AtributeValue is immutable:
https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/dynamodb/model/AttributeValue.html

What about something like this:
```
trait AttributeValue extends js.Object {
  val S: js.UndefOr[String] = js.undefined
  val N: js.UndefOr[String] = js.undefined
  val B: js.UndefOr[Uint8Array] = js.undefined
  val SS: js.UndefOr[js.Array[String]] = js.undefined
  ...
}

object AttributeValue {
  val NULL: AttributeValue = new AttributeValue {
    override val NULL: js.UndefOr[Boolean] = true
  }

  def BOOL(value: Boolean): AttributeValue = new AttributeValue {
    override val BOOL: js.UndefOr[Boolean] = value
  }
  def S(value: String): AttributeValue = new AttributeValue {
    override val S: js.UndefOr[String] = value
  }
  ...
}
```

Reply:
I see, then I think you want to follow your proposal in https://github.com/SystemFw/dynosaur/pull/27#discussion_r998467309 was basically correct.

```scala
@js.native
trait AttributeValue extends js.Object {
  val S: js.UndefOr[String] = js.native
  ...
}
```

Reply:
Cool, thank you!

Reply:
Just to add some colour from the Discord conversation: the reason Dynosaur doesn't define its own ADTs is due to negative past experiences with that approach, and to the fact we end up using the AWS SDK a lot (which is already non blocking and with immutable data), therefore prioritising for interoperability with it by having a wrapper over AttributeValue was both lower effort and more impactful for us.

I'm open to support a smithy4s-backed ADT in the future instead, but only after smithy4s proves to be successful